### ISSUE TITLE
Thoughts about incremental regeneration

### ISSUE THREAD

**Issue by alfredxing** [jekyll/jekyll#4407]
**TL;DR:** Maybe we might be better off approaching incremental regeneration as a cache instead of "incremental regeneration"?

---

So, I've been thinking quite a bit about what I've done with the incremental regeneration feature, and it occurred to me that there might be a better approach.

From the beginning, we've looked at this as "incremental regeneration": regenerating posts and pages only when they have changed. This sounded like a very appealing plan, and it was how I implemented the first (and current) version of incremental regen. It turned out to work okay at first; that is, until we started discovering cases where it would completely break the process.

I'm currently trying to remedy the situation with some fixes and improvements, and these are working fine for now. However, after making my most recent commits in #4269 (and taking a shower :stuck_out_tongue:), I thought about why things turned out to be so complicated.

The biggest problem so far has been how references to other documents' contents are handled. While my aforementioned PR attempts to address some of the issues here, I noticed that even these don't really help out. For example, in the case of an `index.html` containing the standard loop with `{{ post.content }}`, triggering a regeneration of `index.html` because of a change in a post will cause the *untransformed* content of every other post to be included instead. This is because only the post that changed and the index page were re-rendered.

After thinking about it a bit, I realized that perhaps the original thought of "incremental regeneration" wasn't the best idea. So here's a brief outline of my proposal for an alternative:

- Remove most if not all parts of incremental regeneration, and instead create a 'cache'
  - The cache will basically be a shadow version of `_site`, but storing marshalled metadata instead of the final output
- For each page/document, we'll store a marshalled object of the document's rendered output and content (only recently did I realize what the difference between the two was!)
- We render the site normally (like we do without incremental regen)
  - Check to see if the document path exists in the cache index
  - If yes:
    - MD5 hash the original, untransformed content of the document, and compare this to the hash stored in the cache
    - If they match, return the cached content/output
    - If they don't match, that means the content has changed & cache entry invalid; continue to "if no" below
  - If no:
    - Render the document as per normal
    - Create/update the cache entry with the MD5 of the untransformed content, rendered content, and rendered output
- Profit!

We can also use the file mtime instead of the hash like we do now, but this misses the case where the file is saved without changes. Minor edge case; we can investigate the pros/cons of both approaches.

Anyway, I'll probably be playing around with this idea a bit.

I'd love anyone's thoughts or input on this! Am I crazy? Is there a case where this wouldn't work?

@jekyll/core

**Comment by envygeeks** [comment]
My Thoughts:

* MD5 sums can clash easily on large sites (perhaps, it all depends really), SHA is better.
* You'll need a way for things like `jekyll-assets` to tie into that cache to invalidate pages in production.
* `mtime` is unreliable, especially since most programmers never think to copy that too.
* You'll need hefty dependency tracking like Sprockets, *still*

---

* Don't Marshal the meta, make it an open format anybody can audit without Ruby.
* Hold that file in the memory and have Jekyll exit cleaner (if it doesn't already) and write *only* then.
* The cache should be abstract, I like memcached on my servers, you like the FS (maybe.)

**Comment by alfredxing** [comment]
Those are good points!

- Meh, I didn't think the problem would be significant at all; I just said MD5 because I know it's quite fast. Any hash would work though.
- Sure thing, we'll probably provide a method to invalidate by path that just nulls the hash
- Definitely, which is why I proposed the hash first :)
- How so? Using this approach, I was thinking we would be rendering every single document, except in cases where the contents are unchanged we return what's in the cache instead of calling Liquid. Kind of like how a browser's cache works.

---

- On large sites, disk usage will be a con of caching everything, so I was thinking we should choose a format that doesn't take up too much space. Of course it doesn't have to be Marshal; again, any other format would work as well. Gzip and MsgPack could be good candidates.
- Yup!
- Interesting idea. I think leaving this up to plugin developers would be a good idea. We can add in some hooks to the process to enable this.

**Comment by envygeeks** [comment]
> How so? Using this approach, I was thinking we would be rendering every single document, except in cases where the contents are unchanged we return what's in the cache instead of calling Liquid. Kind of like how a browser's cache works.

You have no idea if the contents are going to change unless you either:

1. Let them change (which makes this exercise futile) or
2. You ask all parties involved if they plan to change so you have a valid outlook.

---

If `index.html` relies `default.html` and `default.html` relies on included `head.html` then... if `head.html` changes the entire site needs to be regenerated because all parties rely on `default.html` and `default.html` got regenerated because of `head.html`. This needs cascade backwards. If that doesn't happen, the cache isn't production ready.  If you track `index.html` and it's content only, how exactly do you know if the content changes when content has content of it's own?

Comparing web-caching to this caching is apples to oranges because your browser mostly relies on content that probably already cached upstream in some way, and it will get told when the cache is invalid and update itself, it's just copying, it's not doing the complicated tracking you need to do.  Their caching tacks onto your caching to make both parties seem faster, your caching is making Jekyll more efficient.

**Comment by envygeeks** [comment]
A good example of this need is again `jekyll-assets`, we cache our own assets, we sum our own assets, and in production we offer to attach those sums to the url in some way. If we can't tell Jekyll that `index.html` used `{% img ubuntu.jpg %}` and `ubuntu.jpg` has changed, update yourself with the new url.  Then the caching is broken.

**Comment by envygeeks** [comment]
To make the need for dependency tracking more explainable through `jekyll-assets` take our most common tag `{% css %}` `default.html` has `{% css default %}` if `default.scss` we need to be able to tell Jekyll that default has updated, `default.html` needs to update, which should cascade backwards to `index.html` which relies on `default.html`

**Comment by envygeeks** [comment]
I feel like I'm playing the strawman.  Let me clarify my position: I would like these changes merged into the incremental regenerator and the name changed to just `Cache` because all good caching behaves the way our "incremental regenerator" behaves.  I want to see your changes but I want the cache to be robust like it is headed towards now, I just think incremental regeneration is a bad name to say: "look we do caching... the good way"

**Comment by alfredxing** [comment]
Oops, you're absolutely right! Thanks for explaining. So that was the obvious thing I felt I was missing. :disappointed: 

Okay, so in the end we'll have the cache as described above, but with dependency tracking. Or in other words, incremental regeneration like we have now except with hashes & storing the content/output for each document as well.

With regards to the dependency tracking, do you think what we have right now is okay? Or do you have any ideas for a better way of keeping track of things?

**Comment by parkr** [comment]
@alfredxing Love this issue! Thanks for writing it up in such detail. 

I think the caching idea is an excellent one. Let's get more metrics around what takes lots of time during the build process too, so we can dive into it. If it's rendering, then caching would be very helpful (cache the sources of each page and compare via SHA, regen only what's changed and its dependencies). If it's not rendering, then we know to look elsewhere to optimize.

:smile:
